File: comparison/Haskell/isort-20000.hs
{-
http://www.codecodex.com/wiki/Insertion_sort#Standard_ML
-}
import System.Random

isort [] = []
isort (x:xs) = insert x (isort xs)

insert x [] = [x]
insert x (y:ys) =
  if x<=y then x:y:ys
  else y:(insert x ys)
  

{-
main = print $ isort [3,2,8,5]
-}


randomList :: Int -> IO([Int])
randomList 0 = return []
randomList n = do
  r  <- randomRIO (1,10000)
  rs <- randomList (n-1)
  return (r:rs)


validation [] = True
validation (x:xs) = validation_cons x xs
validation_cons x [] = True
validation_cons x (y:ys) =
  if x<=y then validation_cons y ys
  else False


main = do
  list <- randomList 20000
  let sorted = isort list
  print (validation sorted)

File: comparison/Python/isort-20000.py
import random

def insert_sort(arr):
    for i in range(1, len(arr)):
        j = i - 1
        ele = arr[i]
        while arr[j] > ele and j >= 0:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = ele
    return arr


def mkRandList(n):
    a=[]
    for i in range(1,n+1):
        a.insert(0, random.randint(0,10000))
    return a


def validation(alist):
    for i in range(len(alist)-1):
        if (alist[i] > alist[i+1]):
            return False

    return True


a = mkRandList(20000)
insert_sort(a)
print(validation(a))
